### DISCUSSION TITLE
Proper server data fetching in Client components in NextJS 13

### DISCUSSION THREAD

**Discussion by thefrana** [vercel/next.js#43802] [App Router]
Hi,
I have an issue with the new approach. I use Chakra UI as my component library (awesome by the way). It requires (as other libraries) to use `use client` directive at the top of the pages. My question is how to do server side rendering properly?

The page cannot be async because it is client. If I use SWR, the data is loaded **after** the initial load. I would like the data to be fetched on the server and then passed down to page component. Similar to how it was in Next 12 with `getServerSideProps`. Any help?

Right now, it is unusable. I think it is a problem, because I cannot use the full potential of server components with popular libraries like Chakra or MUI.

Thanks

**Comment by tubbo** [comment]
Server components can mount client components as children:

```typescript
import { Button } from './button'

export default async function Page() {
  const data = await fetch('http://example.com').then((r) => r.json())
  
  return <Button>{JSON.stringify(data)}</Button>
}
```

However, Chakra UI doesn't have `'use client'` directives atop all of its components, so you might need to re-export it from your own component (e.g. **app/button.tsx**):

```typescript
export { Button } from '@chakra-ui/react'
```

Additionally, you'll need to wrap your layout in the `ChakraContext`, which you'll need to do with a client component as well:

```typescript
'use client'

import { ChakraProvider } from '@chakra-ui/react'

export function Config({ children }) {
  return <ChakraProvider>{children}</ChakraProvider>
}
```

inspiration: https://twitter.com/sebmarkbage/status/1587615490616221696

**Reply by thefrana** [reply]
Thank you for answer. It somehow works right now! However, the integration is not great DX. I hope Next.js team will work on that.

**Reply by curiouscod3** [reply]
Cool tip ! The twitter user is working on Vercel btw !
